internal/common/ingest/ingestion_pipeline.go

```diff
@@ -0,0 +1,211 @@
+	MessageIds     []pulsar.MessageID
+}
+
+// IngestionPipeline is an application tha reads message from pulsar and inserts them into a sink. The pipeline will
+// handle the following automatically:
+//   - Receiving messages from pulsar
+//   - Splitting messages into batches for efficient processing
+//   - Unmarshalling into event sequences
+//   - Acking processed messages
+//   - Gracefully terminating on a SIGTERM
+//
+// Callers must supply two structs, an InstructionConverter for converting event sequences into something that can be
+// exhausted and a Sink capable of exhausting these objects
+type IngestionPipeline[T HasPulsarMessageIds] struct {
+	pulsarConfig           configuration.PulsarConfig
+	metricsConfig          configuration.MetricsConfig
+	pulsarSubscriptionName string
+	pulsarbatchSize        int
+	pulsarbatchDuration    time.Duration
+	converter              InstructionConverter[T]
+	sink                   Sink[T]
+}
+
+func NewIngestionPipeline[T HasPulsarMessageIds](pulsarConfig configuration.PulsarConfig,
+	pulsarSubscriptionName string,
+	pulsarbatchSize int,
+	pulsarbatchDuration time.Duration,
+	converter InstructionConverter[T],
+	sink Sink[T],
+	metricsConfig configuration.MetricsConfig,
+) *IngestionPipeline[T] {
+	return &IngestionPipeline[T]{
+		pulsarConfig:           pulsarConfig,
+		metricsConfig:          metricsConfig,
+		pulsarSubscriptionName: pulsarSubscriptionName,
+		pulsarbatchSize:        pulsarbatchSize,
+		pulsarbatchDuration:    pulsarbatchDuration,
+		converter:              converter,
+		sink:                   sink,
+	}
+}
+
+// Run will run the ingestion pipeline until a SIGTERM is received
+func (ingester *IngestionPipeline[T]) Run() {
+	// create a context that will end on a sigterm
+	ctx := createContextWithShutdown()
+
+	shutdownMetricServer := common.ServeMetrics(ingester.metricsConfig.Port)
+	defer shutdownMetricServer()
+
+	// Waitgroup that wil fire when the pipeline has been torn down
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+
+	// Subscribe to Pulsar and receive messages
+	pulsarClient, err := pulsarutils.NewPulsarClient(&ingester.pulsarConfig)
+	if err != nil {
+		panic(errors.WithMessage(err, "Error creating pulsar client"))
+	}
+	defer pulsarClient.Close()
+
+	consumer, err := pulsarClient.Subscribe(pulsar.ConsumerOptions{
+		Topic:                       ingester.pulsarConfig.JobsetEventsTopic,
+		SubscriptionName:            ingester.pulsarSubscriptionName,
+		Type:                        pulsar.KeyShared,
+		SubscriptionInitialPosition: pulsar.SubscriptionPositionEarliest,
+	})
+	defer consumer.Close()
+
+	if err != nil {
+		panic(errors.WithMessage(err, "Error creating pulsar consumer"))
+	}
+	pulsarMsgs := pulsarutils.Receive(ctx, consumer, ingester.pulsarConfig.ReceiveTimeout, ingester.pulsarConfig.BackoffTime)
+
+	// Batch up messages
+	batchedMsgs := Batch(pulsarMsgs, ingester.pulsarbatchSize, ingester.pulsarbatchDuration, clock.RealClock{})
```

Comment:
Here, I think batch should run as a child service in an errgroup. That way cancellation and error propagation just works.